documentation/client_implementation.md

```diff
@@ -57,18 +59,42 @@ type Client interface {
 	// Subscribe creates an ephemeral subscription for the given stream. It
 	// begins receiving messages starting at the configured position and waits
 	// for new messages when it reaches the end of the stream. The default
-	// start position is the end of the stream. It returns an ErrNoSuchStream
-	// if the given stream does not exist. Use a cancelable Context to close a
-	// subscription.
+	// start position is the end of the stream. It returns an
+	// ErrNoSuchPartition if the given stream or partition does not exist. Use
+	// a cancelable Context to close a subscription.
 	Subscribe(ctx context.Context, stream string, handler Handler, opts ...SubscriptionOption) error
 
-	// Publish publishes a new message to the NATS subject. If the AckPolicy is
-	// not NONE and a deadline is provided, this will synchronously block until
-	// the first ack is received. If the ack is not received in time, a
-	// DeadlineExceeded status code is returned. If an AckPolicy and deadline
-	// are configured, this returns the first Ack on success, otherwise it
-	// returns nil.
-	Publish(ctx context.Context, subject string, value []byte, opts ...MessageOption) (*proto.Ack, error)
+	// Publish publishes a new message to the Liftbridge stream. The partition
+	// that gets published to is determined by the provided partition or
+	// Partitioner passed through MessageOptions, if any. If a partition or
+	// Partitioner is not provided, this defaults to the base partition. This
+	// partition determines the underlying NATS subject that gets published to.
+	// To publish directly to a specific NATS subject, use the low-level
+	// PublishToSubject API.
+	//
+	// If the AckPolicy is not NONE and a deadline is provided, this will
+	// synchronously block until the ack is received. If the ack is not
+	// received in time, a DeadlineExceeded status code is returned. If an
+	// AckPolicy and deadline are configured, this returns the Ack on success,
+	// otherwise it returns nil.
+	Publish(ctx context.Context, stream string, value []byte, opts ...MessageOption) (Ack, error)
+
+	// PublishToSubject publishes a new message to the NATS subject. Note that
+	// because this publishes directly to a subject, there may be multiple (or
+	// no) streams that receive the message. As a result, MessageOptions
+	// related to partitioning will be ignored. To publish at the
+	// stream/partition level, use the high-level Publish API.
+	//
+	// If the AckPolicy is not NONE and a deadline is provided, this will
+	// synchronously block until the first ack is received. If an ack is not
+	// received in time, a DeadlineExceeded status code is returned. If an
+	// AckPolicy and deadline are configured, this returns the first Ack on
+	// success, otherwise it returns nil.
+	PublishToSubject(ctx context.Context, subject string, value []byte, opts ...MessageOption) (Ack, error)
```

Comment:
This is awesome! Just to make this understandable for me: does that mean I can transparently add message streams on top of pre-existing NATS subjects? Here's my use case: we use NATS _extensively_ in our apps. We have a couple standard subjects that all follow the same pattern: `object:operation` (e.g. `room:create`, `user:delete`, `company:update` etc.). With this, can I add a stream on top of these so that the current producers don't need to be aware of Liftbridge? And second, can I then use `PublishToSubject` in Liftbridge-aware apps and still have the old consumers listening to the existing NATS subjects be aware of all the messages while still allowing Liftbridge-aware consumers reap the benefits of the durable streams?

Sorry for the long comment – just want to understand the whole machinery. :]

Reply:
Right, you're basically saying exactly that below:
> Liftbridge accepts plain NATS messages, allowing it to make existing subjects durable without any publisher changes. However, these messages will not have features such as acks. […]

This. Is. Awesome. 🎉

Reply:
Yep, I think you got it. :)